_posts/2016-05-25-coordination-in-akka.md

```diff
@@ -0,0 +1,147 @@
+---
+layout: post
+name: Coordination-in-akka
+title: 'Coordination in Akka'
+date: 2016-05-25 00:20:00 +00:00
+author: Felipe Fernández
+image:
+    src: /assets/img/custom/blog/coordination.jpg
+    attribution:
+      text: Castellers! by Nacho Pintos
+      href: https://www.flickr.com/photos/bonus1up/3727496558
+tags:
+- akka
+- coordination
+- supervision
+- akka-http
+---
+
+This is the fifth post in the series about integrating sync clients with async systems ([1, ](http://codurance.com/2016/04/28/async-systems-with-sync-clients/)[2, ](http://codurance.com/2016/04/30/akka-basics/)[3, ](http://codurance.com/2016/05/10/finite-state-machines-with-akka/)[4](http://codurance.com/2016/05/16/publish-subscribe-model-in-kafka/)). Here we'll see how to manage the actor's lifecycle so our service can use efficiently the available resources.
+
+## Lifecycle
+
+Actors, threads, objects, resources... All of them have different states during theirs lifetime. Some of those states are internal and they shape the entity's behaviour when receiving an external stimulus. However there are state transitions that should be handled by external entities: creation, error recovery and disposal.
+
+## Actor Pools
+
+A long time ago dependency injection showed us that an object should not be in charge of creating its dependencies. Externalising those concerns into a central factory makes code more testable, reusable and readable. We'll need different creation strategies disregarding whether we use DI frameworks or poor man's dependency injection. [Guice Scopes](https://github.com/google/guice/wiki/Scopes) is a good example about how to create one instance per application (@Singleton) or one per scope (@SessionScoped and @RequestScoped).
+
+Often the resources are scarce, expensive to create or they're under heavy load. That means that we can't afford nor a Singleton neither an unbounded creation strategy like [Prototype in Spring](http://www.tutorialspoint.com/spring/spring_bean_scopes.htm). A good example are [thread pools](https://docs.oracle.com/javase/tutorial/essential/concurrency/pools.html).
+
+> Using worker threads minimizes the overhead due to thread creation. Thread objects use a significant amount of memory, and in a large-scale application, allocating and deallocating many thread objects creates a significant memory management overhead.
```

Comment:
Not just memory management, takes up CPU bandwidth as well - mention if you feel its relevant